github-api-call-inventory/README.md

```diff
@@ -0,0 +1,37 @@
+# GitHub API Call Rate Limits
+
+## Inventory of GitHub API calls made during a pipeline run
+
+See [this Jupyter Notebook](./github-api-call-inventory.ipynb).
+
+## Collecting metrics on GitHub API calls
+
+We want to count the number of GitHub API calls, so that we can produce an
+indicator about the average API calls per pipeline run. Such an indicator
+could help us notice if code changes result in an increase in the API calls
+made, and so making it more likely that the call rate limits are reached in
+certain orgs.
+
+Packit and Packit-Service interact with GitHub through ogr, which in turn
+relies on PyGithub.
+
+The first thing, that was explored is if PyGithub could be patched in some way
+in order to enable Prometheus metrics collection. This would be actually
+possible, by patching [Requester] to make it possible to accept a Prometheus
+metric object and use it, but I couldn't figure out an easy way to create this
+object in Packit and Packit-Service, pass it to ogr, and then pass it to
+PyGithub.
+
+An another solution that _might_ work is to update the hosts files in the
+service and worker images and route all connections to GitHub through an
+[NGINX proxy pod], and then use [NGINX Prometheus Exporter] to expose some
+[HTTP metrics], which then could be interpreted as metrics on GitHub API
+calls. Exposing these metrics—so that they can be scraped—would require
+setting up a new route, something like `github-calls.packit.dev`. I did not
+try this approach, so some more time should be spent on this, to figure out if
+it would indeed work.
```

Comment:
Would be nice to get a second pair of eyes on this if it indeed makes sense :smiley:

Reply:
Will this "man-in-the-middle" work?

If yes, then it looks like the cleanest solution without any unreliable library patching...

Reply:
I'll try to come up with a small local experiment to try this out.

Reply:
whaaaat, I would love to see this experiment 😍

if it proves reliable, we could use it for other services as well

Reply:
It seems the proxy approach would be more complicated than I assumed. NGINX doesn't support forward proxying out of the box (in this case, this is what is needed, not the reverse proxy feature). Solutions do exist, but they involve patching NGINX code, which I consider quite hacky.

I've started looking at PyGithub again, and searching for a way to inject some response hooks into the `requests` session in `github.Requester`, and figure out a way to reliably pass down these hooks from ogr objects. Metrics collections could happen in these hook functions, and we could also make @TomasTomecek happy and create metrics for (groups of) status codes :slightly_smiling_face: